 # **<font color="DarkBlue">¿Por qué NumPy para arrays (matrices) y computación vectorizada?</font>**

<p align="center">
<img src="https://numpy.org/images/logo.svg" width="80" height="">
</p>


https://numpy.org

<p align="justify">
La computación vectorizada en NumPy permite realizar operaciones matemáticas y de datos en arrays (matrices) de manera eficiente y rápida, evitando el uso de bucles explícitos y aprovechando las optimizaciones de bajo nivel. Esto es crucial para el análisis de datos y la ciencia de datos, ya que mejora la velocidad y eficiencia del procesamiento de grandes volúmenes de datos.

<p align="justify">
📘 NumPy, una abreviatura de "Numerical Python", tiene una historia que se remonta al comienzo del siglo XXI.
<br><br>
✅ <b>Inicios</b>
<br><br>
<b>Orígenes en el Lenguaje Python</b>: A principios de los años 2000, Travis Oliphant, un científico de datos y programador, estaba trabajando en proyectos que involucraban cálculos numéricos intensivos. Sin embargo, encontró que el rendimiento de Python en tales tareas era insuficiente debido a la falta de estructuras de datos eficientes para cálculos numéricos.
<br><br>
<b>Inspiración en MATLAB</b>: Travis se inspiró en la sintaxis y las capacidades de MATLAB, un lenguaje de programación ampliamente utilizado en matemáticas y ciencia de la computación. Buscaba crear un equivalente de código abierto y gratuito que pudiera ofrecer funcionalidades similares a las de MATLAB, pero integradas en Python.
<br><br>
✅ <b>Desarrollo de NumPy</b>
<br><br>
<b>Numerical Python (NumPy) 1.0</b>: En 2006, Travis Oliphant, junto con otros colaboradores, lanzó la versión 1.0 de NumPy. Esta biblioteca ofrecía estructuras de datos como arrays multidimensionales y funciones para operaciones matemáticas sobre los arrays.
<br><br>
<b>Código abierto y colaboración</b>: NumPy fue lanzado como un proyecto de código abierto bajo la licencia BSD, lo que permitió a la comunidad científica y de programación contribuir con mejoras y nuevas características.
<br><br>
✅ <b>Evolución y popularidad</b>
<br><br>
<b>Adopción en la comunidad científica</b>: NumPy rápidamente ganó popularidad entre los científicos de datos, ingenieros y académicos debido a su facilidad de uso y su eficiencia en el manejo de grandes volúmenes de datos numéricos.
<br><br>
<b>Ecosistema científico de Python</b>: La aparición de NumPy impulsó el crecimiento de un ecosistema completo de herramientas y bibliotecas científicas tales como SciPy, Matplotlib, Pandas y Scikit-learn, entre otras.

<p align="justify"> 👀 Por convención, de esta manera se importa <code>Numpy</code>:  </p>

In [ ]:
import numpy as np

 # **<font color="DarkBlue">Introducción a NumPy</font>**

<p align="justify">
📘 NumPy es una biblioteca fundamental en el ecosistema de Python para la computación numérica y científica. Proporciona una manera eficiente de trabajar con arrays multidimensionales (también conocidos como <b>matrices</b>) y funciones matemáticas para realizar operaciones rápidas en estos arrays.
<br><br>
👀 Algunas de las características principales de NumPy incluyen:
<br><br>
<ul align = "justify">
<li>
<b>Eficiencia</b>: NumPy está implementado en C, esto lo hace muy rápido y eficiente en comparación con las listas de Python estándar para operaciones numéricas.</li>
<li>
<b>Arrays Multidimensionales</b>: NumPy proporciona un objeto de array multidimensional llamado <code>ndarray</code>, que permite representar datos en forma de matrices de cualquier número de dimensiones.</li>
<li>
<b>Funciones Matemáticas</b>: NumPy incluye una amplia gama de funciones matemáticas para realizar operaciones numéricas en arrays, como sumas, productos, exponenciación, funciones trigonométricas, y otras funciones.</li>
<li>
<b>Indexación y Slicing Avanzados</b>: NumPy proporciona formas flexibles de acceder y manipular elementos individuales o subconjuntos de elementos en arrays utilizando técnicas de indexación y slicing.</li>
<li>
<b>Álgebra Lineal</b>: NumPy incluye un conjunto completo de funciones para realizar operaciones de álgebra lineal, como la multiplicación de matrices, la inversión de matrices y el cálculo de determinantes y autovalores.
</li></ul>


 # <font color="DarkBlue"><b>Creación de arrays NumPy</font>

<p align="justify">
✅ Creación de Arrays NumPy
<br><br>
Una de las características más poderosas de NumPy es su capacidad para crear y manipular arrays de manera eficiente. Se puede crear arrays NumPy utilizando la función <code>numpy.array()</code> o utilizando otras funciones útiles, como <code>numpy.zeros()</code>, <code>numpy.ones()</code>, <code>numpy.arange()</code> y <code>numpy.linspace()</code>.


## <font color="DarkBlue"><b>Array unidimensional


<p align="center">
<img src="https://github.com/cristiandarioortegayubro/BDS/blob/main/images/array1d.png?raw=true" width="150" height="">
</p>


### <font color="DarkBlue"><b>Creando un array unidimensional


In [ ]:
arr1d = np.array([1, 2, 3, 4, 5])
arr1d

array([1, 2, 3, 4, 5])

In [ ]:
arr1d.shape

(5,)

<p align="justify">
👀 Este script crea un array unidimensional. Básicamente, este código representa la creación y visualización de un array unidimensional utilizando NumPy.

### <font color="DarkBlue"><b>Creando un array de ceros


In [ ]:
zeros_arr1d = np.zeros((9))
zeros_arr1d

array([0., 0., 0., 0., 0., 0., 0., 0., 0.])

<p align="justify">
👀 Este script utiliza la función <code>np.zeros()</code> para crear un array unidimensional de longitud 9, lleno de ceros de manera eficiente.

### <font color="DarkBlue"><b>Creando un array de unos


In [ ]:
ones_arr1d = np.ones((5))
ones_arr1d

array([1., 1., 1., 1., 1.])

<p align="justify">
👀 Este script utiliza la función <code>np.ones()</code> para crear un array unidimensional de longitud 5 de acuerdo al ejemplo, lleno de unos.

### <font color="DarkBlue"><b>Creando un array con valores espaciados uniformemente


In [ ]:
range_arr = np.arange(0, 10, 2)
range_arr

array([0, 2, 4, 6, 8])

<p align = "justify">
👀 Este script utiliza la función <code>np.arange()</code> para crear un array unidimensional que contiene una secuencia de números comenzando desde 0 (inclusive), hasta 10 (exclusivo), con un paso de 2 entre cada número.
<br><br>
A continuación, otros ejemplos:

In [ ]:
range_arr = np.arange(0, 21, 5)
range_arr

array([ 0,  5, 10, 15, 20])

In [ ]:
range_arr = np.arange(0, 31, 3)
range_arr

array([ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27, 30])

### <font color="DarkBlue"><b>Creando un array con valores espaciados linealmente


In [ ]:
linspace_arr = np.linspace(0, 5, 10)
linspace_arr

array([0.        , 0.55555556, 1.11111111, 1.66666667, 2.22222222,
       2.77777778, 3.33333333, 3.88888889, 4.44444444, 5.        ])

<p align="justify">
👀 Este script utiliza la función <code>np.linspace()</code> para crear un array unidimensional que contiene una secuencia de números uniformemente espaciados entre dos valores dados. En este caso, el array comienza en 0 y termina en 5, con un total de 10 elementos igualmente espaciados a lo largo de ese intervalo. Por lo tanto, contendrá valores distribuidos uniformemente entre 0 y 5, incluyendo ambos extremos.

👀 ahora con 5 elementos...

In [ ]:
linspace_arr = np.linspace(0, 5, 5)
linspace_arr

array([0.  , 1.25, 2.5 , 3.75, 5.  ])

👀 si queremos valores enteros, entonces usamos el parámetro <code>dtype</code>

In [ ]:
linspace_arr = np.linspace(0, 5, 5, dtype=int)
linspace_arr

array([0, 1, 2, 3, 5])

## <font color="DarkBlue"><b>Array bidimensional


<p align="center">
<img src="https://github.com/cristiandarioortegayubro/BDS/blob/main/images/array2d.png?raw=true" width="150" height="">
</p>


### <font color="DarkBlue"><b>Creando un array bidimensional



In [ ]:
arr2d = np.array([[1, 2, 3], [4, 5, 6]])
arr2d

array([[1, 2, 3],
       [4, 5, 6]])

<p align="justify">
👀 Este script crea un array bidimensional con una lista de listas que contiene dos filas y tres columnas.
<br><br>
La primera fila contiene los elementos <code>[1, 2, 3]</code> y la segunda fila contiene los elementos <code>[4, 5, 6]</code>. En términos de matriz, esto representa una matriz con dos filas y tres columnas.

In [ ]:
arr2d.shape

(2, 3)

In [ ]:
arr_bidimensional = np.random.random((3, 3))
arr_bidimensional

array([[0.86101255, 0.71009046, 0.07850375],
       [0.02313903, 0.35685505, 0.37517805],
       [0.34489189, 0.84003938, 0.45165833]])

<p align="justify">
👀 Con la función <code>np.random.random()</code> se crea un array. Cada elemento de este arreglo bidimensional es generado aleatoriamente y se encuentra en el rango de 0 a 1, siguiendo una distribución uniforme (distribución de probabilidad en la que todos los valores posibles tienen la misma probabilidad de ocurrir). Por lo tanto representa una matriz de 3 filas y 3 columnas, donde cada elemento es un número aleatorio entre 0 y 1.

### <font color="DarkBlue"><b>Creando un array de ceros



In [ ]:
zeros_arr2d = np.zeros((3, 3))
zeros_arr2d

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [ ]:
zeros_arr2d.shape

(3, 3)

### <font color="DarkBlue"><b>Creando un array de unos



In [ ]:
ones_arr2d = np.ones((2, 2))
ones_arr2d

array([[1., 1.],
       [1., 1.]])

## <font color="DarkBlue"><b>Array n-dimensional


<p align="center">
<img src="https://github.com/cristiandarioortegayubro/BDS/blob/main/images/array3d.png?raw=true" width="180" height="">
</p>


In [ ]:
arr_3d = np.random.random((2, 3, 4))
arr_3d

array([[[0.42406246, 0.17317527, 0.93417046, 0.51466516],
        [0.78473649, 0.88045085, 0.85492746, 0.57868243],
        [0.37603366, 0.67631844, 0.00216592, 0.08404085]],

       [[0.32828421, 0.5513646 , 0.84257416, 0.00206576],
        [0.32996846, 0.47855339, 0.52756007, 0.80791571],
        [0.73811829, 0.69806076, 0.03502208, 0.7486562 ]]])

 # **<font color="DarkBlue">Indexación y Slicing</font>**

<p align="justify">
✅ NumPy permite acceder a elementos individuales, filas, columnas y secciones de arrays utilizando indexación y slicing. Esto proporciona una forma eficiente de manipular y trabajar con datos en arrays NumPy.



 ## **<font color="DarkBlue">Indexación de arrays unidimensionales</font>**

<p align="justify">
Un array unidimensional es esencialmente una lista de elementos. La indexación en este caso es simple y directa

In [ ]:
arr = np.array([0, 1, 2, 3, 4, 5])

In [ ]:
len(arr)

6

 ### **<font color="DarkBlue">Acceso por índice</font>**

<p align="justify">
👀 Los elementos se acceden mediante su índice, que empieza desde 0. Por ejemplo, para un array <code>a = np.array([10, 20, 30, 40, 50])</code>, <code>a[0]</code> devuelve 10, <code>a[1]</code> devuelve 20, y así sucesivamente.

In [ ]:
arr

array([0, 1, 2, 3, 4, 5])

In [ ]:
primer_elemento = arr[0]
primer_elemento

0

 ### **<font color="DarkBlue">Indexación Negativa</font>**

<p align="justify">
También puedes usar índices negativos para acceder a los elementos desde el final del array.

In [ ]:
arr

array([0, 1, 2, 3, 4, 5])

In [ ]:
ultimo_elemento = arr[-1]
ultimo_elemento

5

In [ ]:
arr

array([0, 1, 2, 3, 4, 5])

 ### **<font color="DarkBlue">Subarrays</font>**

<p align="justify">
También puedes acceder a subarrays usando la notación de corte (slicing).

In [ ]:
sub_array = arr[2:5]
sub_array

array([2, 3, 4])

 ## **<font color="DarkBlue">Indexación de arrays bidimensionales</font>**

In [ ]:
arr_2d = np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
arr_2d

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

 ### **<font color="DarkBlue">Acceso por índice</font>**

In [ ]:
elemento = arr_2d[1, 1]
elemento

4

In [ ]:
fila = arr_2d[1]
fila

array([3, 4, 5])

 ### **<font color="DarkBlue">Subarrays</font>**

In [ ]:
columna = arr_2d[:, 1]
columna

array([1, 4, 7])

In [ ]:
seccion = arr_2d[:2, 1:]
seccion

array([[1, 2],
       [4, 5]])

 ## **<font color="DarkBlue">Indexación con expresiones booleanas</font>**

In [ ]:
rand_arr = np.random.random(5)
rand_arr

array([0.47173824, 0.75657576, 0.47222274, 0.77414489, 0.57946013])

In [ ]:
mascara = rand_arr > 0.5
mascara

array([False,  True, False,  True,  True])

In [ ]:
valores_seleccionados = rand_arr[mascara]
valores_seleccionados

array([0.75657576, 0.77414489, 0.57946013])

 # **<font color="DarkBlue">Computación Vectorizada</font>**

<p align="jusfify">
👀 La computación vectorizada en NumPy se refiere a la capacidad de realizar operaciones en arrays completos sin la necesidad de usar bucles explícitos. Esto se logra aprovechando operaciones a nivel de bajo nivel que son altamente optimizadas, haciendo las operaciones más rápidas y eficientes.

 ## **<font color="DarkBlue">Operaciones Básicas de Algebra Lineal</font>**

<p align="justify">
✅ NumPy proporciona un conjunto completo de funciones para realizar operaciones básicas de álgebra lineal en arrays NumPy. Estas operaciones son fundamentales en el análisis de datos y el aprendizaje automático.
<br><br>
A continuación, se presentan algunos ejemplos de operaciones básicas de álgebra lineal:



 ### **<font color="DarkBlue">Transposición de matrices</font>**

<p align="justify">
La transposición de una matriz implica cambiar sus filas por columnas y viceversa. En otras palabras, el elemento en la posición $(i, j)$ de la matriz original se moverá a la posición $(j, i)$ en la matriz transpuesta.
<br><br>
En NumPy, la transposición de matrices se realiza fácilmente usando el método <code>.T</code> o el método <code>transpose()</code>

In [ ]:
matriz = np.array([[1, 2], [3, 4]])
matriz

array([[1, 2],
       [3, 4]])

In [ ]:
transpuesta = np.transpose(matriz)
transpuesta

array([[1, 3],
       [2, 4]])

In [ ]:
transpuesta = transpuesta.T
transpuesta

array([[1, 2],
       [3, 4]])

<p align="justify">
👀 Ejemplo:
<br><br>
Supongamos que estamos analizando las ventas mensuales de diferentes productos en una tienda. Inicialmente, tenemos una matriz donde cada fila representa un producto y cada columna representa un mes.

In [ ]:
# Matriz original: filas son productos y columnas son meses
ventas = np.array([
    [250, 300, 400],   # Producto 1
    [150, 200, 300],   # Producto 2
    [350, 400, 500]])  # Producto 3

print("Matriz original:")
print(ventas)

Matriz original:
[[250 300 400]
 [150 200 300]
 [350 400 500]]


<p align="justify">
La matriz ventas tiene las ventas de tres productos durante tres meses. Ahora, si queremos analizar los datos por meses en lugar de por productos, podemos transponer la matriz...

In [ ]:
# Transposición de la matriz
ventas_transpuesta = ventas.T

print("\nMatriz transpuesta:")
print(ventas_transpuesta)


Matriz transpuesta:
[[250 150 350]
 [300 200 400]
 [400 300 500]]


<p align="justify">
Después de la transposición, cada fila representará un mes y cada columna representará un producto. De esta forma se facilita el análisis de las ventas mensuales.



<p align="justify">
La transposición de matrices en NumPy es una operación útil para reorganizar datos y facilitar el análisis desde diferentes perspectivas. En un contexto de negocios, esto puede ser esencial para generar informes y detectar patrones en los datos

 ### **<font color="DarkBlue">Multiplicación de matrices</font>**

<p align="justify">
La multiplicación de matrices, también conocida como producto matricial, es una operación donde se multiplica cada fila de la primera matriz por cada columna de la segunda matriz y se suman los productos.
<br><br>
En NumPy, esta operación se puede realizar usando el operador <code>@</code> o la función <code>np.dot()</code>.

In [ ]:
matriz1 = np.array([[1, 2], [3, 4]])
matriz1

array([[1, 2],
       [3, 4]])

In [ ]:
matriz2 = np.array([[5, 6], [7, 8]])
matriz2

array([[5, 6],
       [7, 8]])

In [ ]:
producto = np.dot(matriz1, matriz2)
producto

array([[19, 22],
       [43, 50]])

<p align="justify">
👀 Ejemplo:
<br><br>
Supongamos que estamos analizando el impacto de varias campañas de marketing en las ventas de diferentes productos. Tenemos dos matrices:
<br><br>
<b>Matriz de gastos de marketing</b>: Donde cada fila representa un producto y cada columna representa una campaña.
<br><br>
<b>Matriz de efectividad de campañas</b>: Donde cada fila representa una campaña y cada columna representa un impacto esperado en las ventas.
<br><br><br>
El objetivo es calcular la matriz de ventas esperadas, multiplicando la matriz de gastos de marketing por la matriz de efectividad de campañas.

In [ ]:
# Matriz de gastos de marketing: filas son productos, columnas son campañas
gastos_marketing = np.array([
    [100, 200, 150],  # Producto 1
    [80,  160, 120],  # Producto 2
    [50,  100, 75]    # Producto 3
])

In [ ]:
# Matriz de efectividad de campañas: filas son campañas, columnas son impacto en ventas
efectividad_campaña = np.array([
    [1.2, 0.8],  # Campaña 1
    [1.5, 1.0],  # Campaña 2
    [1.3, 0.9]   # Campaña 3
])

In [ ]:
print("Matriz de gastos de marketing:")
print(gastos_marketing)

Matriz de gastos de marketing:
[[100 200 150]
 [ 80 160 120]
 [ 50 100  75]]


In [ ]:
print("\nMatriz de efectividad de campañas:")
print(efectividad_campaña)


Matriz de efectividad de campañas:
[[1.2 0.8]
 [1.5 1. ]
 [1.3 0.9]]


Luego, multiplicamos las dos matrices para obtener la matriz de ventas esperadas:



In [ ]:
# Multiplicación de matrices
ventas_esperadas = gastos_marketing @ efectividad_campaña

print("\nMatriz de ventas esperadas:")
print(ventas_esperadas)


Matriz de ventas esperadas:
[[615.  415. ]
 [492.  332. ]
 [307.5 207.5]]


<p align="justify">
La matriz de ventas esperadas muestra el impacto total de las campañas de marketing en las ventas de cada producto. Por ejemplo, el valor 615 en la primera fila y primera columna indica que el Producto 1 espera tener un incremento total de ventas de 615 unidades (o el valor monetario correspondiente) debido a las campañas de marketing.

<p align="justify">
🏷 En el contexto de negocios, el producto matricial es útil para:
<br><br>

1. **Análisis de ROI**: Calcular el retorno de inversión (ROI) esperado de diferentes campañas de marketing.
2. **Planificación de Estrategias**: Ayudar en la planificación y asignación de recursos a las campañas de marketing más efectivas.

 ### **<font color="DarkBlue">Cálculo de determinantes</font>**

<p align="justify">
El determinante es una propiedad de las matrices cuadradas que proporciona información importante sobre la matriz, como si es invertible o no. Una matriz cuadrada es una matriz en la que el número de filas es igual al número de columnas. Formalmente, una matriz cuadrada de orden \( n \) es una matriz \( A \) de tamaño \( n \times n \) que se puede expresar como:
<br><br><br>
$$ A = \begin{bmatrix}
a_{11} & a_{12} & \cdots & a_{1n} \\
a_{21} & a_{22} & \cdots & a_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
a_{n1} & a_{n2} & \cdots & a_{nn}
\end{bmatrix} $$
<br><br>
donde \( a_{ij} \) representa el elemento en la fila \( i \) y la columna \( j \).
<br><br>
Una matriz invertible, también conocida como matriz no singular o matriz de rango completo, es una matriz cuadrada que tiene una matriz inversa. Formalmente, una matriz cuadrada \( A \) de tamaño \( n \times n \) es invertible si existe otra matriz \( A^{-1} \) tal que:
<br><br>
\[ A \cdot A^{-1} = A^{-1} \cdot A = I \]
<br><br>
donde \( I \) es la matriz identidad de tamaño \( n \times n \). La matriz \( A^{-1} \) se denomina la inversa de \( A \) y satisface la ecuación:
<br><br>
\[ A \cdot A^{-1} = I \]
<br><br>
En NumPy, el determinante de una matriz se calcula utilizando la función <code>np.linalg.det()</code>.

In [ ]:
matriz

array([[1, 2],
       [3, 4]])

In [ ]:
determinante = np.linalg.det(matriz)
determinante

-2.0000000000000004

In [ ]:
determinante.round(2)

-2.0

<p align="justify">
👀 Ejemplo:
<br><br>
Supongamos que estamos analizando la matriz de coeficientes de un sistema de ecuaciones lineales que representa la relación entre diferentes factores económicos (como inversión, producción y ventas) y queremos determinar si este sistema tiene una solución única. El determinante de la matriz de coeficientes puede ayudarnos a entender esto:

- si el determinante es diferente de cero, el sistema tiene una solución única;
- si es cero, el sistema puede tener infinitas soluciones o ninguna.

In [ ]:
# Matriz de coeficientes: filas y columnas representan factores económicos
coeficientes = np.array([
    [2, 3, 1],
    [4, 1, -3],
    [-1, 2, 5]
])

print("Matriz de coeficientes:")
print(coeficientes)

Matriz de coeficientes:
[[ 2  3  1]
 [ 4  1 -3]
 [-1  2  5]]


Ahora, calculamos el determinante de la matriz:

In [ ]:
# Cálculo del determinante
determinante = np.linalg.det(coeficientes)

print("Determinante de la matriz de coeficientes:")
print(determinante.round(2))

Determinante de la matriz de coeficientes:
-20.0


<p align="justify">
El determinante de la matriz de coeficientes es aproximadamente -20. Un determinante diferente de cero indica que el sistema de ecuaciones tiene una solución única, lo que implica que las relaciones entre los factores económicos están bien definidas y el sistema es estable.

 ### **<font color="DarkBlue">Cálculo de inversas</font>**

<p align="justify">
La inversa de una matriz es otra matriz que, cuando se multiplica por la matriz original, da como resultado la matriz identidad.
<br><br>
Una matriz identidad es una matriz cuadrada en la que todos los elementos de la diagonal principal son 1 y todos los elementos fuera de la diagonal principal son 0. Formalmente, una matriz identidad de tamaño \( n \times n \) se denota por \( I_n \) y se define como:
<br><br>
$$ I_n = \begin{bmatrix}
1 & 0 & 0 & \cdots & 0 \\
0 & 1 & 0 & \cdots & 0 \\
0 & 0 & 1 & \cdots & 0 \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
0 & 0 & 0 & \cdots & 1
\end{bmatrix}$$
<br><br>
donde los elementos en la diagonal principal \( i_{ii} \) son 1 y los elementos fuera de la diagonal principal son 0.
<br><br>
En NumPy, la inversa de una matriz se calcula utilizando la función <code>np.linalg.inv()</code>.



In [ ]:
matriz

array([[1, 2],
       [3, 4]])

In [ ]:
inversa = np.linalg.inv(matriz)
inversa

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

<p align="justify">
👀 Ejemplo:
<br><br>
Supongamos que estamos analizando un sistema de ecuaciones lineales que representa la relación entre diferentes productos y su demanda en varios mercados.
<br><br>
Queremos resolver este sistema para encontrar los valores de las variables que representan la cantidad de productos que deben producirse para satisfacer la demanda en cada mercado.



In [ ]:
# Matriz de coeficientes: filas y columnas representan productos y mercados
coeficientes = np.array([
    [4, 2, 1],
    [3, 5, 2],
    [1, 2, 3]
])

print("Matriz de coeficientes:")
print(coeficientes)

Matriz de coeficientes:
[[4 2 1]
 [3 5 2]
 [1 2 3]]


Luego, calculamos la inversa de la matriz:

In [ ]:
# Cálculo de la inversa
inversa = np.linalg.inv(coeficientes)

print("Inversa de la matriz de coeficientes:")
print(inversa)

Inversa de la matriz de coeficientes:
[[ 0.35483871 -0.12903226 -0.03225806]
 [-0.22580645  0.35483871 -0.16129032]
 [ 0.03225806 -0.19354839  0.4516129 ]]


<p align="justify">
Para poner esto en un contexto de negocios, supongamos que tenemos un vector de demanda para cada mercado y queremos encontrar la cantidad de cada producto que debemos producir para satisfacer esta demanda

In [ ]:
# Vector de demanda en cada mercado
demanda = np.array([30, 50, 40])

print("Vector de demanda:")
print(demanda)

Vector de demanda:
[30 50 40]


In [ ]:
# Cantidad de productos a producir para satisfacer la demanda
produccion = inversa @ demanda #multiplicación de matrices

# Redondear los valores a los enteros más cercanos
produccion_enteros = np.round(produccion).astype(int)

print("Cantidad de productos a producir (redondeada a enteros):")
print(produccion_enteros)

Cantidad de productos a producir (redondeada a enteros):
[3 5 9]


<p align="justify">
Con la matriz de coeficientes y el vector de demanda proporcionado, redondeando los resultados obtenidos:
<br><br>

- Se deben producir 3 unidades del primer producto.
- Se deben producir 5 unidades del segundo producto.
- Se deben producir 9 unidades del tercer producto.

 ### **<font color="DarkBlue">Resolución de sistemas de ecuaciones lineales</font>**

<p align="Justify">
Para resolver un sistema de ecuaciones lineales en NumPy, podemos usar la matriz de coeficientes y el vector de constantes para encontrar la solución. El sistema de ecuaciones se puede representar en la forma matricial $Ax = b$, donde:
<br><br>

- $A$ es la matriz de coeficientes.
- $x$ es el vector de variables que queremos encontrar.
- $b$ es el vector de constantes.

<br>
<p align="justify">
La solución se puede encontrar resolviendo $x = A^{-1} b$, pero una manera más eficiente y directa en NumPy es usar la función <code>np.linalg.solve()</code>.
<br><br>
En un sistema de ecuaciones lineales, la <b>matriz de coeficientes</b> es una matriz cuadrada o rectangular que contiene los coeficientes de las variables en las ecuaciones.  Formalmente, para un sistema de \( m \) ecuaciones con \( n \) variables, la matriz de coeficientes \( A \) es una matriz de tamaño \( m \times n \) dada por:
<br><br>
$$ A = \begin{bmatrix}
a_{11} & a_{12} & \cdots & a_{1n} \\
a_{21} & a_{22} & \cdots & a_{2n} \\
\vdots & \vdots & \ddots & \vdots \\
a_{m1} & a_{m2} & \cdots & a_{mn}
\end{bmatrix} $$
<br><br>
donde \( a_{ij} \) representa el coeficiente de la \( i \)-ésima ecuación para la \( j \)-ésima variable.
<br><br>
El <b>vector de constantes</b> es un vector columna que contiene los términos constantes del sistema de ecuaciones. Formalmente, para un sistema de \( m \) ecuaciones, el vector de constantes \( b \) es un vector columna de tamaño \( m \) dado por:
<br><br>
$$ b = \begin{bmatrix}
b_1 \\
b_2 \\
\vdots \\
b_m
\end{bmatrix} $$
<br>
donde \( b_i \) es el término constante en la \( i \)-ésima ecuación.

In [ ]:
A = np.array([[2, 3], [4, 5]])
A

array([[2, 3],
       [4, 5]])

In [ ]:
b = np.array([1, 2])
b

array([1, 2])

In [ ]:
x = np.linalg.solve(A, b)
x

array([0.5, 0. ])

<p align="justify">
👀 Este script utiliza la función <code>np.linalg.solve()</code> para resolver un sistema de ecuaciones lineales representado por la matriz $A$ y el vector $b$. La función <code>np.linalg.solve()</code> encuentra la solución para el vector <code>x</code> que satisface la ecuación $Ax = b$, donde $A$ es una matriz y $b$ es un vector, y se asigna el resultado a la variable $x$.

<p align="justify">
👀 Ejemplo:
<br><br>
Supongamos que tenemos el siguiente sistema de ecuaciones lineales:
<br><br>

- $( 2x + 3y + z = 10 )$
- $( 4x + y - 2z = 5 )$
- $( -x + 2y + 3z = 7 )$

<br><p>
Queremos encontrar los valores de $ x$, $ y $, $ z $.

In [ ]:
# Matriz de coeficientes (A)
coeficientes = np.array([
    [2, 3, 1],
    [4, 1, -2],
    [-1, 2, 3]
])

print("Matriz de coeficientes (A):")
print(coeficientes)

Matriz de coeficientes (A):
[[ 2  3  1]
 [ 4  1 -2]
 [-1  2  3]]


In [ ]:
# Vector de constantes (b)
constantes = np.array([10, 5, 7])

print("Vector de constantes (b):")
print(constantes)

Vector de constantes (b):
[10  5  7]


In [ ]:
# Resolución del sistema de ecuaciones lineales
solucion = np.linalg.solve(coeficientes, constantes)

# Redondear los valores a enteros más cercanos
solucion_enteros = np.round(solucion).astype(int)

print("Solución del sistema de ecuaciones (redondeada a enteros):")
print(solucion_enteros)

Solución del sistema de ecuaciones (redondeada a enteros):
[2 1 2]


<p align="justify">
Después de redondear los resultados:
<br><br>

- $x = 2$
- $y = 1$
- $z = 2$

<br>
<p align="justify">
Estos valores son las aproximaciones enteras de la solución del sistema de ecuaciones. Ten en cuenta que el redondeo puede llevar a una solución que no satisface exactamente las ecuaciones originales, pero puede ser útil para simplificar la interpretación en algunos casos.

 ## **<font color="DarkBlue">Otras operaciones de Computación Vectorizada</font>**

<p align="justify">
Entonces, en NumPy la computación vectorizada se refiere a la realización de operaciones matemáticas en arrays de manera eficiente, sin la necesidad de usar bucles explícitos.
<br><br>
A continuación las principales operaciones de computación vectorizada en NumPy y una breve explicación de cada una:

 ### **<font color="DarkBlue">Operaciones Aritméticas Elementales</font>**

Estas operaciones se aplican a cada elemento de un arreglo individualmente:
- **Suma** (`+`): Suma los elementos de dos arreglos o de un arreglo con un escalar.
- **Resta** (`-`): Resta los elementos de un arreglo de los elementos de otro arreglo o de un escalar.
- **Multiplicación** (`*`): Multiplica los elementos de dos arreglos o de un arreglo por un escalar.
- **División** (`/`): Divide los elementos de un arreglo entre los elementos de otro arreglo o entre un arreglo y un escalar.


 ### **<font color="DarkBlue">Funciones Matemáticas</font>**

NumPy proporciona una serie de funciones matemáticas que se aplican elemento por elemento:
- **Raíz cuadrada** (`np.sqrt`): Calcula la raíz cuadrada de cada elemento.
- **Exponencial** (`np.exp`): Calcula el exponencial (e elevado a la potencia) de cada elemento.
- **Logaritmo** (`np.log`): Calcula el logaritmo natural de cada elemento.
- **Trigonometría** (`np.sin`, `np.cos`, `np.tan`): Calcula las funciones trigonométricas de cada elemento.



 ### **<font color="DarkBlue">Operaciones de Agregación</font>**


Estas operaciones resumen los datos de un arreglo:
- **Suma total** (`np.sum`): Suma todos los elementos del arreglo.
- **Promedio** (`np.mean`): Calcula el promedio de todos los elementos.
- **Mínimo y Máximo** (`np.min`, `np.max`): Encuentra el valor mínimo y máximo, respectivamente.



 ### **<font color="DarkBlue">Operaciones de Comparación</font>**


Estas operaciones devuelven un arreglo de valores booleanos:
- **Igualdad** (`==`): Compara si los elementos de dos arreglos son iguales.
- **Desigualdad** (`!=`): Compara si los elementos de dos arreglos son diferentes.
- **Mayor que / Menor que** (`>`, `<`): Compara si los elementos de un arreglo son mayores o menores que los elementos de otro arreglo o un escalar.



 ### **<font color="DarkBlue">Operaciones de Transformación de Datos</font>**


Estas operaciones cambian la forma o el formato de los arreglos:
- **Redimensionamiento** (`np.reshape`): Cambia la forma de un arreglo sin cambiar sus datos.
- **Transposición** (`np.transpose`): Cambia la orientación de las dimensiones de un arreglo.



<br>
<br>
<p align="center"><b>
💗
<font color="DarkBlue">
Hemos llegado al final de nuestro colab, a seguir codeando en NumPy...
</font>
</p>
